text/0000-name-resolution.md

```diff
@@ -0,0 +1,649 @@
+
+
+#### The order of declarations of items should be irrelevant.
+
+I.e., names should be able to be used before they are declared. Note that this
+clearly does not hold for declarations of variables in statements inside
+function bodies.
+
+
+#### Macros should be manually expandable.
+
+Compiling a program should have the same result before and after expanding a
+macro 'by hand', so long as hygiene is accounted for.
+
+
+#### Glob imports should be manually expandable.
+
+A programmer should be able to replace a glob import with a list import that
+imports any names imported by the glob and used in the current scope, without
+changing name resolution behaviour.
+
+
+#### Visibility should not affect name resolution.
+
+Clearly, visibility affects whether a name can be used or not. However, it
+should not affect the mechanics of name resolution. I.e., changing a name from
+public to private (or vice versa), should not cause more or fewer name
+resolution errors (it may of course cause more or fewer accessibility errors).
+
+
+## Changes to name resolution rules
+
+### Multiple unused imports
+
+A name may be imported multiple times, it is only a name resolution error if
+that name is used. E.g.,
+
+```
+mod foo {
+    pub struct Qux;
+}
+
+mod bar {
+    pub struct Qux;
+}
+
+mod baz {
+    use foo::*;
+    use bar::*; // Ok, no name conflict.
+}
+```
+
+In this example, adding a use of `Qux` in `baz` would cause a name resolution
+error.
+
+### Multiple imports of the same binding
+
+A name may be imported multiple times and used if both names bind to the same
+item. E.g.,
+
+```
+mod foo {
+    pub struct Qux;
+}
+
+mod bar {
+    pub use foo::Qux;
+}
+
+mod baz {
+    use foo::*;
+    use bar::*;
+
+    fn f(q: Qux) {}
+}
+```
+
+### non-public imports
+
+Currently `use` and `pub use` items are treated differently. Non-public imports
+will be treated in the same way as public imports, so they may be referenced
+from modules which have access to them. E.g.,
+
+```
+mod foo {
+    pub struct Qux;
+}
+
+mod bar {
+    use foo::Qux;
+
+    mod baz {
+        use bar::Qux; // Ok
+    }
+}
+```
+
+
+### Glob imports of accessible but not public names
+
+Glob imports will import all accessible names, not just public ones. E.g.,
+
+```
+struct Qux;
+
+mod foo {
+    use super::*;
+
+    fn f(q: Qux) {} // Ok
+}
+```
+
+This change is backwards incompatible. However, the second rule above should
+address most cases, e.g.,
+
+```
+struct Qux;
+
+mod foo {
+    use super::*;
+    use super::Qux; // Legal due to the second rule above.
+
+    fn f(q: Qux) {} // Ok
+}
+```
+
+The below rule (though more controversial) should make this change entirely
+backwards compatible.
+
+Note that in combination with the above rule, this means non-public imports are
+imported by globs where they are private but accessible.
+
+
+### Explicit names may shadow implicit names
+
+Here, an implicit name means a name imported via a glob or inherited from an
```

Comment:
What wins in case of conflict, a glob or an inherited name?
This currently compiles and prints `glob`. Seems reasonable.
```
mod m {
    pub struct S;
    impl S { pub fn print(&self) -> &'static str { "glob" } }
}

fn main() {
    struct S;
    impl S { pub fn print(&self) -> &'static str { "inherited" } }
    {
        use m::*;
        println!("{}", S.print());
    }
}
```